### PR TITLE
[SPARK-16283][SQL] Implement `percentile_approx` SQL function

### PR DESCRIPTION
**Pull request by lw-lin**
## What changes were proposed in this pull request?

This patch Implements `percentile_approx` SQL function using Spark's implementation of G-K algorithm.
- commit 1: moves the G-K algorithm implementation(`QuantileSummaries` and related tests) from `sql/core` to `sql/catalyst`
- commit 2: implements `percentile_approx` using G-K algorithm
## How was this patch tested?
- Jenkins
- added new tests

### CODE REVIEW

**Comment by lw-lin** [comment]
Jenkins retest this please

**Comment by lw-lin** [comment]
@hvanhovell could you take a look at this? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+    case Literal(i: Int, IntegerType) if i > 0 => i
+
+    case _ =>
+      throw new AnalysisException("The third argument should be a positive integer literal")
+  }
+}
+
+/**
+ * Helper class to compute approximate quantile summary.
+ * This implementation is based on the algorithm proposed in the paper:
```
Comment (thunterdb) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: why did you need to move the implementation around?

Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: thanks!

module `sql/core` depends on module `sql/catalyst`, but module `sql/catalyst` does not depend on module `sql/core`; so in order to enable expressions like `PercentileApprox` (which lives in `sql/catalyst`) to use of the G-K algorithm, the algorithm should probably live in `sql/catalyst`. As a reference, algorithms like HyperLogLog also live in `sql/catalyst`(please refer to [HyperLogLogPlusPlus](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala)).

@thunterdb does this make sense?

**Comment by lw-lin** [comment]
@cloud-fan could you also help review this? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala (hunk 1 above)
Comment (thunterdb) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: Yes, this makes sense, thanks.

**Comment by lw-lin** [comment]
Jenkins retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+  // Please refer to SPARK-16283 for details.
+  private[sql] def bToRelativeError(B: Int): Double = Math.max(1.0d / B, 0.001)
+
+  /**
+   * Validates the percentile(s) expression and extract the percentile(s).
+   * Returns the extracted percentile(s) and an indicator of whether it's an array.
+   */
+  private def validatePercentilesLiteral(exp: Expression): (Seq[Double], Boolean) = {
+    def withinRange(v: Double): Boolean = 0.0 <= v && v <= 1.0
+    exp match {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: You could also pattern match on the result of the `exp.eval()`. That would be way easier.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+import org.apache.spark.sql.catalyst.expressions.aggregate.QuantileSummaries.Stats
+import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.types._
+
+/**
+ * Computes an approximate percentile (quantile) using the G-K algorithm (see below), for very
+ * large numbers of rows where the regular percentile() UDAF might run out of memory.
+ *
+ * The input is a single double value or an array of double values representing the percentiles
+ * requested. The output, corresponding to the input, is either an single double value or an
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: minor: `an single` -> `a single`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+     memory; higher values yield better approximations.
+    _FUNC_(col, array(p1 [, p2]...) [, B]) - Same as above, but accepts and returns an array of
+     percentile values instead of a single one.
+    """)
+case class PercentileApprox(
+    child: Expression,
+    percentilesExpr: Expression,
+    bExpr: Option[Expression],
+    percentiles: Seq[Double],  // the extracted percentiles
+    B: Int,                    // the extracted B
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: Hi, @lw-lin .
I know the reason why you define this as a capatal 'B', but I'm just wondering it's consistent with Spark naming rule.

Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: I don't have strong preference here -- let's see what reviewers say.

**Comment by lw-lin** [comment]
@hvanhovell comments addressed. Please let me know when there's more to do.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, AnyDataType, IntegralType)
+
+  override def checkInputDataTypes(): TypeCheckResult =
+    TypeUtils.checkForNumericExpr(child.dataType, "function percentile_approx")
+
+  // The number of intermediate outputs is highly relative to the actual data-set (an upper bound is
+  // (11/2e)log(2en), where e is the relativeError parameter, n is the number of items in the
+  // dataset) -- thus it's hard to allocate agg buffer in advance without knowing the size of
+  // inputs. Due to this reason, currently we don't support partial mode.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: can you explain a bit more about this? AFAIK, hive supports partial aggregate for `percentile_approx`, and it looks to me that your implementation keeps the buffer data(`QuantileSummaries`) in this aggregate function object, instead of letting aggregate operator manage it, that's the main reason why we can't support partial aggregate for `percentile_approx` I think.

Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: > hive supports partial aggregate for percentile_approx

Hive's implementation computes approximate percentile values from a histogram, thus Hive supports partial aggregation but makes no approximation guarantees.

> ... QuantileSummaries in this aggregate function object, instead of letting aggregate operator manage it, that's the main reason why we can't support partial aggregate

Yes that's quite right. `QuantileSummaries` has been implemented and well tested prior to this patch, so it'd be great if we can directly put a `QuantileSummaries` instance into the aggregation buffer (in order to support partial aggregation). @cloud-fan any pointer on how to do that?

Thanks!

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: to be clear, are you saying that our current algorithm to compute `percentile_approx` can't support partial aggregation fundamentally?

Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: Fundamentally the `QuantileSummaries` implementation does support RDD-style partial aggregation -- `QuantileSummaries` itself is the agg buffer of a partition's data at mappers, and multiple `QuantileSummaries`s will be merged at reducers (please refer to [StatFunctions.multipleApproxQuantiles](https://github.com/apache/spark/blob/5a3533e779d8e43ce0980203dfd3cbe343cc7d0a/sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala#L93)).

Fundamentally it should also support the SparkSQL-style partial aggregation. I'm trying to reuse `QuantileSummaries`; if there's no easy way to reuse `QuantileSummaries`, I'm afraid we'll have to re-write this GK algorithm totally to support our SparkSQL-style partial aggregation

So any way we can just directly put a `QuantileSummaries` instance into SparkSQL's agg buffer? Or do we have to break a `QuantileSummaries` instance up, say into `an int + a double + a double[] + an int[] + an int[]` which SparkSQL's agg buffer can manage?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: I think putting buffer object in agg buffer row is better, but that need to be well designed.  Can you hold this PR for a while? We are discussing about it internally.

Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: Sure, this can wait. Thanks for the information!

Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: Some updates, https://github.com/apache/spark/pull/14753
is created to support putting generic object in aggregation buffer.

Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: @clockfly  cool!
now that https://github.com/apache/spark/pull/14754 has been merged, looks like I should update this patch once https://github.com/apache/spark/pull/14753 gets merged.

Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala]: @liancheng @cloud-fan should I update this or do you guys want to take this over?

**Comment by clockfly** [comment]
@lw-lin 
There are other issues to fix besides partial aggregation, like out of memory, argument type (B, and percentages should use catalyst expression instead so that we support constant folding), naming, input type check, sql. I have created another PR, and acknowledged your work in the PR description. Do you want to take a look at https://github.com/apache/spark/pull/14868?

**Comment by lw-lin** [comment]
@clockfly 
Your PR's improvements overall look reasonable to me, so I'm closing this one. Good job!
